hastory-cli/test/Hastory/Cli/OptParseSpec.hs

```diff
@@ -0,0 +1,393 @@
+                  { remoteStorageClientInfoBaseUrl = url
+                  , remoteStorageClientInfoUsername = username
+                  , remoteStorageClientInfoPassword = password
+                  }
+          Instructions dispatch _settings <-
+            combineToInstructions cmd flags emptyEnvironment emptyConfiguration
+          dispatch `shouldBe`
+            DispatchGenGatherWrapperScript
+              GenGatherWrapperScriptSettings {genGatherWrapperScriptSetRemoteInfo = Just remoteInfo}
+      context "Environment contains all 3 fields of a RemoteStorageClientInfo" $
+        it "is DispatchGenGatherWrapperScript with remote storage info" $ do
+          url <- parseBaseUrl "api.example.com"
+          username <- parseUsername "hastory"
+          let cmd = CommandGenGatherWrapperScript GenGatherWrapperScriptFlags
+              env =
+                emptyEnvironment
+                  { envCacheDir = Nothing
+                  , envStorageServer = Just url
+                  , envStorageUsername = Just username
+                  , envStoragePassword = Just password
+                  }
+              password = "Passw0rd"
+              remoteInfo =
+                RemoteStorageClientInfo
+                  { remoteStorageClientInfoBaseUrl = url
+                  , remoteStorageClientInfoUsername = username
+                  , remoteStorageClientInfoPassword = password
+                  }
+          Instructions dispatch _settings <-
+            combineToInstructions cmd emptyFlags env emptyConfiguration
+          dispatch `shouldBe`
+            DispatchGenGatherWrapperScript
+              GenGatherWrapperScriptSettings {genGatherWrapperScriptSetRemoteInfo = Just remoteInfo}
+      context "Configuration contains all 3 fields of a RemoteStorageClientInfo" $
+        it "is DispatchGenGatherWrapperScript with remote storage info" $ do
+          url <- parseBaseUrl "api.example.com"
+          username <- parseUsername "hastory"
+          let cmd = CommandGenGatherWrapperScript GenGatherWrapperScriptFlags
+              conf =
+                emptyConfiguration
+                  { configCacheDir = Nothing
+                  , configStorageServer = Just url
+                  , configStorageUsername = Just username
+                  , configStoragePassword = Just password
+                  }
+              password = "Passw0rd"
+              remoteInfo =
+                RemoteStorageClientInfo
+                  { remoteStorageClientInfoBaseUrl = url
+                  , remoteStorageClientInfoUsername = username
+                  , remoteStorageClientInfoPassword = password
+                  }
+          Instructions dispatch _settings <-
+            combineToInstructions cmd emptyFlags emptyEnvironment conf
+          dispatch `shouldBe`
+            DispatchGenGatherWrapperScript
+              GenGatherWrapperScriptSettings {genGatherWrapperScriptSetRemoteInfo = Just remoteInfo}
+
+emptyFlags :: Flags
```

Comment:
You also shouldn't need empty flags or an empty configuration. Best to remove them and see what breaks.

Reply:
Is it OK to leave these in for testing only? Otherwise, I will need to build values over and over again with only one field a `Just` and the rest `Nothing`s. I'm perfectly fine w/ taking them out if you still want me to do so, but just wanted to point out that they are only in the spec files. Please let me know.

Reply:
Ah yes if it's only used in the tests then that's fine. Thanks for pointing that out!